apps/api-documenter/src/documenters/YamlDocumenter.ts

```diff
@@ -554,33 +606,50 @@ export class YamlDocumenter {
    * Example:  node-core-library.JsonFile.load
    */
   protected _getUid(apiItem: ApiItem): string {
-    let result: string = '';
-    for (const hierarchyItem of apiItem.getHierarchy()) {
+    return this._getUidObject(apiItem).toString();
+  }
 
-      // For overloaded methods, add a suffix such as "MyClass.myMethod_2".
-      let qualifiedName: string = hierarchyItem.displayName;
-      if (ApiParameterListMixin.isBaseClassOf(hierarchyItem)) {
-        if (hierarchyItem.overloadIndex > 1) {
-          // Subtract one for compatibility with earlier releases of API Documenter.
-          // (This will get revamped when we fix GitHub issue #1308)
-          qualifiedName += `_${hierarchyItem.overloadIndex - 1}`;
-        }
-      }
+  private _getUidObject(apiItem: ApiItem): DeclarationReference {
+    if (apiItem instanceof ApiDeclaredItem && apiItem.uid !== undefined) {
+      return apiItem.uid;
+    }
 
+    let result: DeclarationReference = new DeclarationReference();
```

Comment:
While debugging, I am finding that `UidGenerator._symbolToUid()` returns `undefined` for many of the package exports.  A ton of .api.json entries are missing the `uid` field.  Some of it is just bugs that can be fixed.  But fundamentally `UidGenerator._symbolToUid()` walks backwards from the symbol, trying to reach a package entry point, and returning `undefined` if not.  For an exported symbol, I find it difficult to prove that this will always produce a UID.  And even if it does, the UID may disagree with the path that API Extractor follows to reach the `AstDeclaration`.  This disagreement would occur because API Extractor's analysis crawls forward from the entry point, and thus aliases may be traversed along a different path.

The code here in `_getUidObject()` looks like it compensates for the missing UIDs by constructing them from the .api.json model.  But this seems hacky; every item in the .api.json should have a UID.

Thus I'm thinking it might be safer to do this:

1. For exported APIs, always calculate their UID based on API Extractor's model
2. Do it in **api-extractor-model** instead of here in **api-documenter**
3. For excerpt links, first check whether we can match the symbol to something that was crawled by the `Collector` and already has a known UID; this is the idea that @adventure-yunfei used in https://github.com/microsoft/web-build-tools/pull/1005
4. Otherwise, we could use `UidGenerator._symbolToUid()` as a heuristic to make a best attempt.  However I also notice that `_symbolToUid()` is going to produce a lot of UIDs that don't correspond to any DocFX page.  I wonder if that is going to have a nontrivial performance cost.

@rbuckton Unless you have any objections, I'm going to take a shot at doing (1) and (2) as part of this PR, since seeing the correct UID for every crawled item is important for test validation of our spec.  Whereas (3) and (4) just need to be roughly correct; if some hyperlinks are missing, those are minor bugs that we can fix over time.

Reply:
The main reason to construct a UID from a Symbol is to get an accurate canonical ID for a reference to a declaration/type that is not included in your package's .api.json (such as from an external package). This is so that if the documentation for the two packages were to be merged, or if the external package provided an *xrefmap.yml* file for its documentation, that the UIDs would align and the documentation links would be seamless.

Reply:
@rbuckton That makes sense.  However for the APIs that /are/ part of the .api.json, it's much simpler and more reliable to calculate them from the model.  So far I haven't encountered any cases where information is lost compared to the other approach.  (Well, ECMAScript symbols are not handled correctly, but that's merely an incomplete feature that should be solved by improving **api-extractor-model**.)

I peeled this work off into a separate [PR #1406](https://github.com/microsoft/web-build-tools/pull/1406).  Could you take a look at it and let me know if it looks right?  I am mostly interested in whether I used your class correctly, and whether the generated UIDs match your design.  I'll add some specific questions as comments to that PR.